### PR TITLE
feat: scheduler (6/): add scheduler profile + plugin interfaces

### PR DESCRIPTION
**Pull request by michaelawyu**
### Description of your changes

This PR is part of the PRs that implement the Fleet workload scheduling.

It features the implementation for scheduler profile + the plugin interfaces to the scheduler.

I have:

- [x] Run `make reviewable` to ensure this PR is ready for review.

### How has this code been tested

- [x] Unit tests

### CODE REVIEW

File: pkg/scheduler/framework/profile_test.go [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+	profile.WithPreScorePlugin(dummyAllPurposePlugin)
+	profile.WithScorePlugin(dummyAllPurposePlugin)
+
+	// Same plugin should be registered only once, even if it connects to multiple extension points.
+	if len(profile.registeredPlugins) != 1 {
+		t.Fatalf("registerPlugins len() = %d, want %d", len(profile.registeredPlugins), 1)
+	}
+	if pl, ok := profile.registeredPlugins[dummyPluginName]; !ok || pl != dummyPlugin {
+		t.Fatalf("registeredPlugins[%s] = %v, %t, want %v, %t", dummyPluginName, pl, ok, dummyPlugin, true)
+	}
```
Comment (zhiying-lin) [inline pkg/scheduler/framework/profile_test.go:72]: could you please compare the structure directly?

eg,
want := map[string]Plugin{
		dummyPluginName: dummyPlugin,
	}

	if !cmp.Equal(profile.registeredPlugins, want) {
	}

File: pkg/scheduler/framework/profile_test.go [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+	profile.WithFilterPlugin(dummyAllPurposePlugin)
+	profile.WithPreScorePlugin(dummyAllPurposePlugin)
+	profile.WithScorePlugin(dummyAllPurposePlugin)
+
+	// Same plugin should be registered only once, even if it connects to multiple extension points.
+	if len(profile.registeredPlugins) != 1 {
+		t.Fatalf("registerPlugins len() = %d, want %d", len(profile.registeredPlugins), 1)
+	}
+	if pl, ok := profile.registeredPlugins[dummyPluginName]; !ok || pl != dummyPlugin {
+		t.Fatalf("registeredPlugins[%s] = %v, %t, want %v, %t", dummyPluginName, pl, ok, dummyPlugin, true)
```
Comment (zhiying-lin) [inline pkg/scheduler/framework/profile_test.go:71]: prefer to use errorf in this case,

"Even after your test cases encounter a failure, they should keep going for as long as possible in order to print out all of the failed checks in a single run. This way, someone who's fixing the failing test doesn't have to play whac-a-mole, fixing one bug and then re-running the test to find the next bug.

On a practical level, prefer calling t.Error over t.Fatal. When comparing several different properties of a function's output, use t.Error for each of those comparisions.
"

https://github.com/golang/go/wiki/TestComments#keep-going

File: pkg/scheduler/framework/profile_test.go (hunk 1 above)
Comment (michaelawyu) [reply pkg/scheduler/framework/profile_test.go:72]: Done.

File: pkg/scheduler/framework/profile_test.go (hunk 2 above)
Comment (michaelawyu) [reply pkg/scheduler/framework/profile_test.go:71]: No longer applicable.

**Review by zhiying-lin — APPROVED** [review]

**Pull request merged**